[SPARK-18356] [ML] Improve MLKmeans Performance

ZakariaHili opened the pull request:
## What changes were proposed in this pull request?

Spark Kmeans fit() doesn't cache the RDD which generates a lot of warnings : 
 WARN KMeans: The input data is not directly cached, which may hurt performance if its parent RDDs are also uncached.
So, Kmeans should cache the internal rdd before calling the Mllib.Kmeans algo, this helped to improve spark kmeans performance by 14%

https://github.com/ZakariaHili/spark/commit/a9cf905cf7dbd50eeb9a8b4f891f2f41ea672472

@hhbyyh 
## How was this patch tested?
Pass Kmeans tests and existing tests

srowen commented on mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala:
```diff
@@ -233,7 +232,10 @@ object KMeansModel extends MLReadable[KMeansModel] {
       val dataPath = new Path(path, "data").toString
 
-      val clusterCenters = if (majorVersion(metadata.sparkVersion) >= 2) {
+      val versionRegex = "([0-9]+)\\.(.+)".r
```
Your change is reverting some other changes like this one. Can you clean this up to only reflect the changes you intend?

ZakariaHili replied:
sorry for my late, I was busy,
now , I think the code is ready

AmplabJenkins commented:
Can one of the admins verify this patch?

hhbyyh commented on mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala:
```diff
@@ -33,11 +33,10 @@ import org.apache.spark.rdd.RDD
 /**
  * Common params for KMeans and KMeansModel
- */
+ */ 
```
remove the extra space, please use [Spark root]/dev/lint-scala to verify your code style.

ZakariaHili replied:
I cleaned the code, then I run /dev/lint-scala and I received : 
Scalastyle checks passed.

hhbyyh commented on mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala:
```diff
@@ -304,14 +306,26 @@ class KMeans @Since("1.5.0") (
+    /*
+      Before fit the dataset, we check storage level of the rdd
+      if rdd is not cached, we cached our RDD
+    */
```
The comment may not be necessary, as it's used elsewhere and clear by the code itself

hhbyyh commented:
@ZakariaHili Great to see your PR. @srowen is right, please try to merge with the latest code.

ZakariaHili reviewed (commented):
sorry for my late reply, i was a little bit busy.
I cleaned the code and I tested the code style using ./dev/lint-scala.

hhbyyh commented:
Hi @ZakariaHili is it finished? https://github.com/apache/spark/pull/15965/files still contains a lot of unrelated code change.

ZakariaHili commented:
Hi Yuhao , are you sure ?
I think that you still compare previous commits with

AmplabJenkins commented:
Can one of the admins verify this patch?

ZakariaHili committed:
[SPARK-18356] [ML] Improve MLKmeans Performance

AmplabJenkins commented:
Can one of the admins verify this patch?

ZakariaHili commented:
Sorry, I didn't know that you already accepted some changes from other people, so I had to force the synchronization, then re-test everything.
Now, the code is clean

srowen commented on mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala:
```diff
@@ -306,12 +307,19 @@ class KMeans @Since("1.5.0") (
+    val handlePersistence = dataset.rdd.getStorageLevel == StorageLevel.NONE
+    fit(dataset, handlePersistence)
+  }
+  @Since("2.0.0")
```
This still has several problems. This can't be since 2.0.0 and there is no space before this line.

ZakariaHili replied:
Ok it's removed

srowen commented on mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala:
```diff
@@ -321,12 +329,15 @@ class KMeans @Since("1.5.0") (
     model.setSummary(summary)
     instr.logSuccess(model)
+    if (handlePersistence) {
+     instances.unpersist()
```
fix spacing here

ZakariaHili replied:
the space is fixed

ZakariaHili committed:
[SPARK-18356] [ML] Improve MLKmeans Performance

ZakariaHili reviewed (commented):
@srowen,  I added the space and I removed @ Since("2.0.0")

srowen commented on mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala:
```diff
@@ -306,12 +307,18 @@ class KMeans @Since("1.5.0") (
+    val handlePersistence = dataset.rdd.getStorageLevel == StorageLevel.NONE
+    fit(dataset, handlePersistence)
+  }
+  protected def fit(dataset: Dataset[_], handlePersistence: Boolean): KMeansModel = {
```
Still wrong -- no space, and now no `@Since` at all. This is `@Since("2.2.0")` if it goes in.

ZakariaHili committed:
[SPARK-18356] [ML] Improve MLKmeans Performance

ZakariaHili commented on mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala:
```diff
@@ -305,11 +305,12 @@ class KMeans @Since("1.5.0") (
     fit(dataset, handlePersistence)
-  }
+}
+  @Since("2.2.0")
```
@srowen ,now I think I understand what do you mean by no space here.

ZakariaHili reviewed (commented):
@srowen I hope that i understood you right, this time

srowen commented on mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala:
```diff
@@ -304,14 +305,21 @@ class KMeans @Since("1.5.0") (
   def setSeed(value: Long): this.type = set(seed, value)
 
-  @Since("2.0.0")
+  @Since("2.2.0")
```
Why did you change this? this is wrong.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala:
```diff
@@ -304,14 +305,21 @@ class KMeans @Since("1.5.0") (
+    val handlePersistence = dataset.rdd.getStorageLevel == StorageLevel.NONE
+    fit(dataset, handlePersistence)
+}
+  @Since("2.2.0")
```
There is still no blank line above this line -- not sure how else to express this?
The indentation of the line above is off too.

Look, you should match the style you see around this code, and you can see how it isn't the same? just make it match.

ZakariaHili committed:
[SPARK-18356] [ML] Improve MLKmeans Performance

ZakariaHili commented:
@srowen, Ok I found it
I fixed

ZakariaHili commented:
@srowen, I think it's ready to merge

srowen commented:
I think that's OK. @hhbyyh ?

hhbyyh commented:
LGTM.

SparkQA commented:
**[Test build #3436 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3436/consoleFull)** for PR 15965 at commit [`fd4543d`](https://github.com/apache/spark/commit/fd4543d918b26b451000fe92fbe1cdb259a6687b).

SparkQA commented:
**[Test build #3436 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3436/consoleFull)** for PR 15965 at commit [`fd4543d`](https://github.com/apache/spark/commit/fd4543d918b26b451000fe92fbe1cdb259a6687b).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

ZakariaHili committed:
Update KMeans.scala

ZakariaHili commented on mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala:
```diff
@@ -310,7 +310,7 @@ class KMeans @Since("1.5.0") (
     fit(dataset, handlePersistence)
   }
-  
+
```
I didn't see that with vim editor

SparkQA commented:
**[Test build #3438 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3438/consoleFull)** for PR 15965 at commit [`f17a54c`](https://github.com/apache/spark/commit/f17a54c34e9aa007f36426c2b8459ab0bf784863).

SparkQA commented:
**[Test build #3438 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3438/consoleFull)** for PR 15965 at commit [`f17a54c`](https://github.com/apache/spark/commit/f17a54c34e9aa007f36426c2b8459ab0bf784863).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala:
```diff
@@ -306,12 +307,20 @@ class KMeans @Since("1.5.0") (
+  }
+
+  @Since("2.2.0")
+  protected def fit(dataset: Dataset[_], handlePersistence: Boolean): KMeansModel = {
```
Hi, sorry for not seeing this earlier.  We should not make a new public method for this.  @ZakariaHili would you mind sending a follow-up PR (under the same JIRA issue) which eliminates this new fit() method?  You should be able to delete lines 311 - 315 here and keep everything else the same.  Thank you!

ZakariaHili replied:
Hi @jkbradley, 
In actually, I thought of the same thing, but Mr. yuhao yang advised me to do the same style of LogisticRegression.scala
anywhere I changed the file as you want, and I changed the code a little (e.g : no need to generate the rdd two times
the first one to check storage level and the second one for Kmeans
)
here the change
https://github.com/ZakariaHili/spark/commit/58de5499d0930e4019c94718a2a9ed3dd715acb2

do you still want from me to send a PR ?

jkbradley replied:
I like the change in terms of eliminating the extra fit() method (which is not done anywhere else in spark.ml), but I think you'll need to keep the original test for storage level ```val handlePersistence = dataset.rdd.getStorageLevel == StorageLevel.NONE```.  In the current change, the RDD being tested is different from ```dataset.rdd```, so handlePersistence would always be true regardless of the original storage level.  Could you please send a little PR which merges the 2 fit() methods?  Thanks!

ZakariaHili commented:
Hi @jkbradley, 
In actually, I thought of the same thing, but Mr. yuhao yang advised me to do the same style of LogisticRegression.scala
anywhere I changed the file as you want, and I changed the code a little (e.g : no need to generate the rdd two time
the first one to check storage level and the second one for Kmeans
)
